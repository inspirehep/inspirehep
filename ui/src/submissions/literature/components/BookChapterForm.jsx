import React, { Component } from 'react';
import { Form, Field } from 'formik';
import { Row } from 'antd';
import PropTypes from 'prop-types';

import CollapsableForm from '../../common/components/CollapsableForm';
import BasicInfoFields from './BasicInfoFields';
import SubmitButton from '../../common/components/SubmitButton';
import LinkFields from './LinkFields';
import ReferencesField from './ReferencesField';
import CommentsField from './CommentsField';
import TextField from '../../common/components/TextField';
import SuggesterField from '../../common/components/SuggesterField';
import BookSuggestion from './BookSuggestion';

const OPEN_SECTIONS = ['basic_info', 'links', 'publication_info'];

class BookChapterForm extends Component {
  static getSuggestionSourceFirstTitle(suggestion) {
    return suggestion._source.titles[0].title;
  }

  static renderBookSuggestion(suggestion) {
    const book = suggestion._source;
    return <BookSuggestion book={book} />;
  }

  render() {
    const { values } = this.props;
    return (
      <Form>
        <CollapsableForm openSections={OPEN_SECTIONS}>
          <CollapsableForm.Section header="Links" key="links">
            <LinkFields />
          </CollapsableForm.Section>
          <CollapsableForm.Section header="Basic Info" key="basic_info">
            <BasicInfoFields values={values} />
          </CollapsableForm.Section>
          <CollapsableForm.Section
            header="Publication Info"
            key="publication_info"
          >
            <Field
              name="book_title"
              recordFieldPath="parent_book_record"
              label="Book Title"
              pidType="literature"
              suggesterName="book_title"
              extractItemCompletionValue={
                BookChapterForm.getSuggestionSourceFirstTitle
              }
              renderResultItem={BookChapterForm.renderBookSuggestion}
              component={SuggesterField}
            />
            <Field name="start_page" label="Start Page" component={TextField} />
            <Field name="end_page" label="End Page" component={TextField} />
          </CollapsableForm.Section>
          <CollapsableForm.Section header="References" key="references">
            <ReferencesField values={values} />
          </CollapsableForm.Section>
          <CollapsableForm.Section header="Comments" key="comments">
            <CommentsField values={values} />
          </CollapsableForm.Section>
        </CollapsableForm>
        <Row type="flex" justify="end">
          <SubmitButton />
        </Row>
      </Form>
    );
  }
}

BookChapterForm.propTypes = {
  isSubmitting: PropTypes.bool.isRequired,
  isValidating: PropTypes.bool.isRequired,
  isValid: PropTypes.bool.isRequired,
  values: PropTypes.objectOf(PropTypes.any).isRequired, // current form data
};

export default BookChapterForm;
