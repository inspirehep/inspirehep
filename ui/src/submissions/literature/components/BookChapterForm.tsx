import React, { Component } from 'react';
import { Form, Field } from 'formik';
import { Row } from 'antd';
import PropTypes from 'prop-types';

import CollapsableForm from '../../common/components/CollapsableForm';
import BasicInfoFields from './BasicInfoFields';
import SubmitButton from '../../common/components/SubmitButton';
import LinkFields from './LinkFields';
import ReferencesField from './ReferencesField';
import CommentsField from './CommentsField';
import TextField from '../../common/components/TextField';
import SuggesterField from '../../common/components/SuggesterField';
import BookSuggestion from './BookSuggestion';
import { Suggestion } from '../../../types';

const OPEN_SECTIONS = ['basic_info', 'links', 'publication_info'];

function BookChapterForm({
  values,
  isSubmitting,
  isValidating,
  isValid,
}: {
  values: any;
  isSubmitting?: boolean;
  isValidating?: boolean;
  isValid?: boolean;
}) {
  function getSuggestionSourceFirstTitle(suggestion: Suggestion) {
    return suggestion._source.titles[0].title;
  }

  function renderBookSuggestion(suggestion: Suggestion) {
    const book = suggestion._source;
    return <BookSuggestion book={book} />;
  }

  return (
    <Form>
      <CollapsableForm openSections={OPEN_SECTIONS}>
        <CollapsableForm.Section header="Links" key="links">
          <LinkFields />
        </CollapsableForm.Section>
        <CollapsableForm.Section header="Basic Info" key="basic_info">
          <BasicInfoFields values={values} />
        </CollapsableForm.Section>
        <CollapsableForm.Section
          header="Publication Info"
          key="publication_info"
        >
          <Field
            name="book_title"
            recordFieldPath="parent_book_record"
            label="Book Title"
            pidType="literature"
            suggesterName="book_title"
            extractItemCompletionValue={getSuggestionSourceFirstTitle}
            renderResultItem={renderBookSuggestion}
            component={SuggesterField}
          />
          <Field name="start_page" label="Start Page" component={TextField} />
          <Field name="end_page" label="End Page" component={TextField} />
        </CollapsableForm.Section>
        <CollapsableForm.Section header="References" key="references">
          <ReferencesField values={values} />
        </CollapsableForm.Section>
        <CollapsableForm.Section header="Comments" key="comments">
          <CommentsField values={values} />
        </CollapsableForm.Section>
      </CollapsableForm>
      <Row justify="end">
        <SubmitButton />
      </Row>
    </Form>
  );
}

export default BookChapterForm;
