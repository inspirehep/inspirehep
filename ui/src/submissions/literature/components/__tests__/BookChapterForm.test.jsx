import React from 'react';
import { render } from '@testing-library/react';
import { Formik } from 'formik';

import BookChapterForm from '../BookChapterForm';

describe('BookChapterForm', () => {
  it('renders all props', () => {
    const { asFragment } = render(
      <Formik>
        <BookChapterForm
          isSubmitting={false}
          isValidating={false}
          isValid
          values={{}}
        />
      </Formik>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
