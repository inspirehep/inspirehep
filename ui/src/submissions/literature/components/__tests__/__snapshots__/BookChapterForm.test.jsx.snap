// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookChapterForm renders all props 1`] = `
<Form>
  <CollapsableForm
    openSections={
      Array [
        "basic_info",
        "links",
        "publication_info",
      ]
    }
  >
    <CollapsableFormSection
      header="Links"
      key="links"
    >
      <LinkFields />
    </CollapsableFormSection>
    <CollapsableFormSection
      header="Basic Info"
      key="basic_info"
    >
      <BasicInfoFields
        values={Object {}}
        withCollaborationField={false}
      />
    </CollapsableFormSection>
    <CollapsableFormSection
      header="Publication Info"
      key="publication_info"
    >
      <Field
        component={[Function]}
        extractItemCompletionValue={[Function]}
        label="Book Title"
        name="book_title"
        pidType="literature"
        recordFieldPath="parent_book_record"
        renderResultItem={[Function]}
        suggesterName="book_title"
      />
      <Field
        component={[Function]}
        label="Start Page"
        name="start_page"
      />
      <Field
        component={[Function]}
        label="End Page"
        name="end_page"
      />
    </CollapsableFormSection>
    <CollapsableFormSection
      header="References"
      key="references"
    >
      <ReferencesField
        values={Object {}}
      />
    </CollapsableFormSection>
    <CollapsableFormSection
      header="Comments"
      key="comments"
    >
      <CommentsField
        values={Object {}}
      />
    </CollapsableFormSection>
  </CollapsableForm>
  <Row
    gutter={0}
    justify="end"
    type="flex"
  >
    <SubmitButton
      isSubmitting={false}
      isValid={true}
      isValidating={false}
    />
  </Row>
</Form>
`;
